docker-compose.yaml

```diff
@@ -19,79 +19,126 @@ x-default: &node
+    image: ${CCAO_REGISTRY_URL}/service-spark-iasworld:latest
+    container_name: spark-node-master-prod
     hostname: spark-node-master
+    restart: unless-stopped
     environment:
-      - SPARK_MODE=master
-
-      - AWS_SHARED_CREDENTIALS_FILE=/run/secrets/AWS_CREDENTIALS
-      - AWS_S3_BUCKET
-      - AWS_S3_PREFIX
-      - SPARK_LOCAL_STORAGE_ENCRYPTION_ENABLED=no
-      - SPARK_PUBLIC_DNS=${SPARK_PUBLIC_IP}
-      - SPARK_RPC_AUTHENTICATION_ENABLED=no
-      - SPARK_RPC_ENCRYPTION_ENABLED=no
-      - SPARK_SSL_ENABLED=no
-      - SPARK_USER=spark
-      - IPTS_PRD_SERVICE_NAME
-      - IPTS_PRD_HOSTNAME
-      - IPTS_PRD_USERNAME
-      - IPTS_PRD_PORT
-      - IPTS_TST_SERVICE_NAME
-      - IPTS_TST_HOSTNAME
-      - IPTS_TST_USERNAME
-      - IPTS_TST_PORT
-      - GH_APP_ID
+      <<: [*spark-env]
+      SPARK_MODE: master
+      SPARK_UI_PORT: 4040
+      SPARK_EXECUTOR_MEMORY: 96g
+      SPARK_MASTER_WEBUI_PORT: 8080
     ports:
       - 4040:4040
       - 8080:8080
+    networks:
+      - sparknet-prod
 
-  spark-node-worker:
+  spark-node-worker-prod:
     <<: *node
-    container_name: spark-node-worker
+    profiles: ["prod", ""]
+    image: ${CCAO_REGISTRY_URL}/service-spark-iasworld:latest
+    container_name: spark-node-worker-prod
     hostname: spark-node-worker
+    restart: unless-stopped
+    depends_on:
+      - spark-node-master-prod
     environment:
-      - SPARK_MODE=worker
-      - SPARK_MASTER_URL=spark://spark-node-master:7077
-      - SPARK_WORKER_MEMORY=96G
-      - SPARK_WORKER_CORES=28
-      - SPARK_WORKER_WEBUI_PORT=9090
-
-      - AWS_SHARED_CREDENTIALS_FILE=/run/secrets/AWS_CREDENTIALS
-      - AWS_S3_BUCKET
-      - AWS_S3_PREFIX
-      - SPARK_PUBLIC_DNS=${SPARK_PUBLIC_IP}
-      - SPARK_RPC_AUTHENTICATION_ENABLED=no
-      - SPARK_RPC_ENCRYPTION_ENABLED=no
-      - SPARK_LOCAL_STORAGE_ENCRYPTION_ENABLED=no
-      - SPARK_SSL_ENABLED=no
-      - SPARK_USER=spark
-      - IPTS_PRD_SERVICE_NAME
-      - IPTS_PRD_HOSTNAME
-      - IPTS_PRD_USERNAME
-      - IPTS_PRD_PORT
-      - IPTS_TST_SERVICE_NAME
-      - IPTS_TST_HOSTNAME
-      - IPTS_TST_USERNAME
-      - IPTS_TST_PORT
-      - GH_APP_ID
+      <<: [*spark-env]
+      SPARK_MODE: worker
+      SPARK_WORKER_MEMORY: 96G
+      SPARK_WORKER_CORES: 32
+      SPARK_WORKER_WEBUI_PORT: 9090
     ports:
       - 9090:9090
+    networks:
+      - sparknet-prod
+
+  spark-node-master-dev:
+    <<: *node
+    profiles: ["dev"]
+    image: ${CCAO_REGISTRY_URL}/service-spark-iasworld:dev
+    container_name: spark-node-master-dev
+    hostname: spark-node-master
+    restart: no
+    environment:
+      <<: [*spark-env]
+      SPARK_MODE: master
+      SPARK_UI_PORT: 4042
+      SPARK_EXECUTOR_MEMORY: 48g
+      # 8081 is the worker port by default and binding it will block jobs. So
+      # we skip to 8082 instead of using the incremented port of 8081
+      SPARK_MASTER_WEBUI_PORT: 8082
```

Comment:
**[Question, non-blocking]** Do we actually need to change the port numbers that we bind within the container? My understanding is that the container's ports are fully isolated until we publish them to host ports using the `ports` service config, so there's no risk of port collision until we publish them to a host port. If that's so, can't we continue binding e.g. `SPARK_MASTER_WEBUI_PORT` to `8080` in the containers for both environments, and then just make sure to publish `8082:8080` in the `ports` config for the dev service?

Reply:
You're right that we probably don't need to change them, but IMO it's better to play it safe here since Spark is pretty finnicky re: networking.